[Refactoring]: Introduce CallRouter

marcomaddy1982 opened the pull request:
## What's new in this PR?

This PR will cover what mentioned here in the [RFC](https://wearezeta.atlassian.net/wiki/spaces/IOS/pages/124125940/Call+Routing+Refactoring).

Main `CallRouter` actions:
- Present/Dismiss/Minimize `ActiveCallController`
- Show and hidde `CallTopOverlayController`
- Present/Dismiss `CallQualityViewController`
- Present `Call Alerts`

Here the two protocols implemented in the `CallRouter`

```swift
// MARK: - CallRouterProtocol
protocol CallRouterProtocol: class {
    func presentActiveCall(for voiceChannel: VoiceChannel, animated: Bool)
    func dismissActiveCall(animated: Bool, completion: (()-> Void)?)
    func minimizeCall(animated: Bool, completion: (() -> Void)?)
    func showCallTopOverlayController(for conversation: ZMConversation)
    func hideCallTopOverlayController()
    func presentSecurityDegradedAlert(degradedUser: UserType?)
    func presentUnsupportedVersionAlert()
}

// MARK: - CallQualityRouterProtocol
protocol CallQualityRouterProtocol: class {
    func presentSurvey(with callDuration: TimeInterval)
    func dismissCurrentSurveyIfNeeded()
    func callQualityControllerDidScheduleDebugAlert()
}
```

P.s: cause this PR is already huge I will add the tests in a separate PR.

David-Henner commented on Wire-iOS/Sources/UserInterface/Overlay/ActiveCallViewController.swift:
```diff
@@ -24,14 +24,14 @@ import WireSyncEngine
 fileprivate let zmLog = ZMSLog(tag: "calling")
 
+protocol ActiveCallViewControllerDelegate: class {
+    func callControllerDidDisappear(_ callController: CallViewController)
```
isn't it redundant to define two different protocols with the same exact method?

David-Henner replied:
the first one was defined in `CallViewController.swift`

David-Henner commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallRouter.swift:
```diff
@@ -0,0 +1,217 @@
+// MARK: - CallRouterProtocol
+protocol CallRouterProtocol: class {
+    func presentActiveCall(for voiceChannel: VoiceChannel, animated: Bool)
+    func dismissActiveCall(animated: Bool, completion: (()-> Void)?)
```
Do we have a type alias for `()-> Void` we could use ?

marcomaddy1982 replied:
I can have a look, if yes I will use it.

David-Henner commented on Wire-iOS/Sources/UserInterface/Calling/CallViewController/CallViewController.swift:
```diff
@@ -149,13 +153,13 @@ final class CallViewController: UIViewController {
 
         if isInteractiveDismissal {
-            dismisser?.dismiss(viewController: self, completion: nil)
+            delegate?.callControllerDidDisappear(self)
```
I don't understand.
`CallViewControllerDelegate` seems to report the disappearance of `self`
`ViewControllerDismisser` seems to ask for dismissal of `self`

how can one replace the other?

marcomaddy1982 replied:
You are right `ViewControllerDismisser` ask for dismissal of `self` but in the old implementation the dismisser was not doing any action of dismissing (moreover the `CallController` is not a `ViewController`). There were just reset some properties (see old implementation). The dismiss was basically called at the end of the swipe down of the active call.

```swift
func dismiss(viewController: UIViewController, completion: Completion? = nil) {
        guard let callViewController = viewController as? CallViewController,
            let conversation = callViewController.conversation else { return }

        minimizedCall = conversation
        activeCallViewController = nil
}
```

My idea is to have a delegate the once the active call is out of the screen reset the property and dismiss the controller.

```swift
func callControllerDidDisappear(_ callController: CallViewController) {
        router?.dismissActiveCall(animated: true, completion: nil)
        minimizedCall = callController.conversation
    }
```

But if you thing that this is confusing or the naming is confusing we can try to improve this part.

David-Henner replied:
Thanks for clarifying! The naming is good 👍

David-Henner reviewed (commented):
Thanks for all the heavy work! 👍

marcomaddy1982 commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallController.swift:
```diff
@@ -21,230 +21,135 @@ import WireSyncEngine
+        guard let userSession = ZMUserSession.shared() else { return }
+        guard let priorityCallConversation = userSession.priorityCallConversation else { dismissCall(); return }
+        
+        topOverlayCall = priorityCallConversation
```
that's a good point!! Thanks. Yes, probably it is not the scope of this PR but I guess we could try to improve it now that we working with this part of the code. Thanks for pointing it out

marcomaddy1982 commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallRouter.swift:
```diff
@@ -0,0 +1,217 @@
+    }
+    
+    func dismissActiveCall(animated: Bool = true, completion: (()-> Void)? = nil) {
+        rootViewController.dismiss(animated: animated, completion: { [weak self] in
```
When I developed this part at the beginning it was done as your comment, then I opted for this approach (that works btw) cause I didn't want to have an instance to the ActiveCallController around in a way that the CallRouter was not too huge. Anyway not that you commented I'm rethinking again. So if I understood you would prefer the first approach, maybe using a lazy property?

```swift
struct Animal {
    private lazy var activeCallViewController: ActiveCallViewControllerDelegate
}
```

David-Henner replied:
I thought at first that it was a mistake. Maybe the naming is a bit confusing, because I'm not sure if `rootViewController` is the root VC of the app or only the starting point for calling related UI.

marcomaddy1982 replied:
Yes, right. But the `rootViewController` of type `RootViewController`. We have just one instance of the `RootViewController` the tit is the root of the app.

marcomaddy1982 commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallRouter.swift:
```diff
@@ -0,0 +1,217 @@
+    }
+    
+    func presentCallQualityRejection() {
+        DebugAlert.showSendLogsMessage(message: "Sending the debug logs can help us improve the quality of calls and the overall app experience.")
```
I could be that I miss it. Also if not it would not hurt adding `executeOrSchedulePostCallAction`. Thanks for pointing it out!!!

marcomaddy1982 commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallRouter.swift:
```diff
@@ -0,0 +1,217 @@
+    }
+}
+
+// MARK: - UIViewControllerTransitioningDelegate
+extension CallRouter: UIViewControllerTransitioningDelegate {
+    func animationController(forPresented presented: UIViewController,
+                             presenting: UIViewController,
+                             source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
+        return (presented is CallQualityViewController) ? CallQualityPresentationTransition() : nil
+    }
+    
+    func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
+        return (dismissed is CallQualityViewController) ? CallQualityDismissalTransition() : nil
+    }
+}
```
I could create a `CallQualityTransitioningDelegate` class that implements `UIViewControllerTransitioningDelegate`, what do you think about?

David-Henner replied:
Yeah I think it'd be great to have in a separate class 👍 
But shouldn't the `delegate` keyword be reserved for protocols only?

marcomaddy1982 replied:
Good point. Maybe `CallQualityAnimator`? What do you think about?

marcomaddy1982 commented on Wire-iOS/Sources/UserInterface/Calling/CallViewController/CallViewController.swift:
```diff
@@ -20,9 +20,13 @@ import UIKit
 import avs
 
+protocol CallViewControllerDelegate: class {
+    func callControllerDidDisappear(_ callController: CallViewController)
```
nice catch!!! Thanks

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallController.swift:
```diff
@@ -21,230 +21,140 @@ import WireSyncEngine
     
-    private func setupCallQualityController() {
-        callQualityController.delegate = self
+    // MARK: - Public Impletation
```
Typo "Implementation"

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallController.swift:
```diff
@@ -21,230 +21,140 @@ import WireSyncEngine
-    private func handleDegradedConversationIfNeeded(_ conversation: ZMConversation) {
-        guard let degradationState = conversation.voiceChannel?.degradationState else {
-            return
+    private func shouldAnimate(call: ZMConversation) -> Bool {
```
Suggestion: `shouldAnimateTransitionForCall(in conversation: ZMConversation) -> Bool`

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallRouter.swift:
```diff
@@ -0,0 +1,215 @@
+import WireSyncEngine
+
+// MARK: - CallRouterProtocol
+protocol CallRouterProtocol: class {
```
Suggestion `ActiveCallRouterProtocol`

marcomaddy1982 replied:
Yes I know, thanks for remembering. We discussed it in the RFC. My idea was apply this change in the next push :-)

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallRouter.swift:
```diff
@@ -0,0 +1,215 @@
+}
+
+// MARK: - CallRouter
+class CallRouter: NSObject {
```
Suggestion: `ActiveCallRouter`

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Calling/CallController/CallRouter.swift:
```diff
@@ -0,0 +1,215 @@
+    }
+        
+    // MARK: - Public Implementation
+    func updateCallState() {
+        callController.updateState()
+    }
```
I think a more intent-revealing name is needed here.

marcomaddy1982 replied:
`updateActiveCallPresentetionState` maybe?

johnxnguyen replied:
Yup, I like that (without the typo of course 😄 )

marcomaddy1982 replied:
right!!! :-D

johnxnguyen commented on Wire-iOS/Sources/UserInterface/Calling/CallController/ActiveCallRouter.swift:
```diff
@@ -0,0 +1,215 @@
+    }
+        
+    // MARK: - Public Implementation
+    func updateCallState() {
+        callController.updateActiveCallPresentetionState()
+    }
```
You should rename `updateCallState` to `updateActiveCallPresentationState` too.

marcomaddy1982 commented:
@David-Henner @johnxnguyen: I have addressed all your comments. It would be nice if you will spend 5 minutes when you have time to try out if no bugs are introduced in the routing.  It is a pretty important part of the app and I could have missed something. Thanks :-)

johnxnguyen commented:
> It would be nice if you will spend 5 minutes when you have time to try out if no bugs are introduced in the routing.

@marcomaddy1982 sure, but it might need to wait until tomorrow unfortunately.

marcomaddy1982 commented:
> > It would be nice if you will spend 5 minutes when you have time to try out if no bugs are introduced in the routing.
> 
> @marcomaddy1982 sure, but it might need to wait until tomorrow unfortunately.

@johnxnguyen it is not a problem. Thanks 👍

David-Henner commented:
> @David-Henner @johnxnguyen: I have addressed all your comments. It would be nice if you will spend 5 minutes when you have time to try out if no bugs are introduced in the routing. It is a pretty important part of the app and I could have missed something. Thanks :-)

I'll try to review again today, but if not, I'll make sure it's done tomorrow

marcomaddy1982 commented:
> > @David-Henner @johnxnguyen: I have addressed all your comments. It would be nice if you will spend 5 minutes when you have time to try out if no bugs are introduced in the routing. It is a pretty important part of the app and I could have missed something. Thanks :-)
> 
> I'll try to review again today, but if not, I'll make sure it's done tomorrow

No rush. We just started the new sprint and this ticket is part of that Sprint.

David-Henner commented on Wire-iOS/Sources/UserInterface/Calling/CallController/ActiveCallRouter.swift:
```diff
@@ -0,0 +1,216 @@
+protocol ActiveCallRouterProtocol: class {
+    func presentActiveCall(for voiceChannel: VoiceChannel, animated: Bool)
+    func dismissActiveCall(animated: Bool, completion: Completion?)
+    func minimizeCall(animated: Bool, completion: (() -> Void)?)
```
Suggestion: `(() -> Void)?` -> `Completion?`

netbe commented on wire-ios/WireUITests/Pages/ActiveConversationPage.swift:
```diff
@@ -188,16 +188,10 @@ class ActiveConversationPage: PageModel {
-            }
+        messageLabels.allElementsBoundByIndex.compactMap { element in
+            let text = element.label.isEmpty ? element.value as? String : element.label
+            return text?.replacingOccurrences(of: "\u{00A0}", with: " ")
```
question: what's this value?

findms replied:
sent message value based on indeitifer

netbe commented on wire-ios/WireUITests/Pages/ActiveConversationPage.swift:
```diff
@@ -188,16 +188,10 @@ class ActiveConversationPage: PageModel {
-                messages.append(normalized)
-            }
+        messageLabels.allElementsBoundByIndex.compactMap { element in
+            let text = element.label.isEmpty ? element.value as? String : element.label
```
question: why is it the value or label?

findms replied:
As per debug description we are reading the value from the identifier
`TextView, 0x107b09ee0, {{149.3, 347.3}, {224.7, 19.3}}, identifier: 'message', value: hello! 17786567341...`
Also reverted this as previously working function..not explicitly needed to change this as this was failing occasionally due to view not loaded after sending message due to too fast which fixed with adding an assert in test.

netbe commented on wire-ios/WireUITests/Pages/UserProfilePage.swift:
```diff
@@ -39,15 +39,15 @@ class UserProfilePage: PageModel {
     var teamNameOnAccountPage: XCUIElement {
-        app.descendants(matching: .any)[Locators.UserProfilePage.teamName.rawValue].firstMatch
+        app.descendants(matching: .staticText)[Locators.UserProfilePage.teamName.rawValue].firstMatch
     }
 
     var manageTeamButton: XCUIElement {
         app.buttons[Locators.UserProfilePage.addAccountOrTeamButton.rawValue].firstMatch
     }
 
     var closeButton: XCUIElement {
-        app.descendants(matching: .any)[Locators.UserProfilePage.close.rawValue].firstMatch
+        app.descendants(matching: .button)[Locators.UserProfilePage.close.rawValue].firstMatch
```
question: is it really necessary to specify the type?

findms replied:
Not necessary but to avoid failures where for some elements it returns like statictext, textview and button..so keep it specific to the type.

netbe replied:
I don't get it, is .any flaky?
